### PR TITLE
sd-daemon: add notify_on_cleanup() helper function and use it where applicable

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/libsystemd/meson.build [hunk 1]
```diff
@@ -90,7 +90,7 @@ libsystemd_sources = files('''
         sd-path/sd-path.c
         sd-resolve/sd-resolve.c
         sd-utf8/sd-utf8.c
-'''.split()) + id128_sources + sd_daemon_c + sd_event_c + sd_login_c
+'''.split()) + id128_sources + sd_daemon_c + sd_event_sources + sd_login_c
```
Comment (keszybz) [inline src/libsystemd/meson.build:93]: The other ones should be renamed at some point too...

File: src/timesync/timesyncd.c [hunk 2]
```diff
@@ -83,7 +85,8 @@ static int load_clock_timestamp(uid_t uid, gid_t gid) {
         return 0;
 }
 
-int main(int argc, char *argv[]) {
+static int run(int argc, char *argv[]) {
+        DEFINE_DEFAULT_NOTIFY_MESSAGE_ON_EXIT;
```
Comment (keszybz) [inline src/timesync/timesyncd.c]: This means that the cleanup action (sending `STOPPING=1...`) will be performed e.g. if argument parsing fails. I don't think we want this.

File: src/timesync/timesyncd.c [hunk 3]
```diff
@@ -149,27 +143,24 @@ int main(int argc, char *argv[]) {
...
-                log_error_errno(r, "Failed to parse fallback server strings: %m");
-                goto finish;
-        }
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse fallback server strings: %m");
 
         log_debug("systemd-timesyncd running as pid " PID_FMT, getpid_cached());
-        sd_notify(false,
-                  "READY=1\n"
-                  "STATUS=Daemon is running");
```
Comment (keszybz) [inline src/timesync/timesyncd.c:160]: A different approach: define
```C
#define NOTIFY_READY "READY=1\n" "STATUS=Daemon is running"
#define NOTIFY_STOPPING  "STOPPING=1\n" "STATUS=Shutting down..."
const char *notify_start(const char *start, const char *stop) {
        sd_notify(false, start);
        return stop;
}
```

And then just wrap part of code like this one in
```C
{
         _cleanup_(notify_on_cleanup) const char *status = notify_start(NOTIFY_READY, NOTIFY_STOPPING);

         r = sd_event_loop(m->event);
         if (r < 0)
                return ...;
}
```
If it is necessary to override either the starting or stopping message, they can be override in the arguments to `notify_start`, and the stop message can also be overridden by setting `status`.

File: src/network/wait-online/wait-online.c [hunk 4]
```diff
@@ -105,7 +110,9 @@ static int parse_argv(int argc, char *argv[]) {
         return 1;
 }
 
-int main(int argc, char *argv[]) {
+static int run(int argc, char *argv[]) {
+        _cleanup_(notify_on_exit) const char *notify_message_on_failure =
+                "STATUS=Failed waiting for network connectivity...";
```
Comment (keszybz) [inline src/network/wait-online/wait-online.c]: That's just too much magic, with the failure code pre-configured and overwritten on success.

File: src/libsystemd/sd-daemon/daemon-util.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#pragma once
+
+#include <stdbool.h>
+
+#include "sd-daemon.h"
+
+/* This is intended to be used with _cleanup_ attribute. */
+static inline void notify_on_exit(const char **p) {
```
Comment (keszybz) [inline src/libsystemd/sd-daemon/daemon-util.h]: Please rename this to `notify_on_cleanup`, or maybe even better `notify_cleanupp`, becuase it's not inherently tied to exiting.

File: src/network/wait-online/wait-online.c (hunk 4 above)
Comment (yuwata) [reply src/network/wait-online/wait-online.c]: Yeah, I think this is tricky :-)

**Comment by yuwata** [comment]
@keszybz Thank you for the review. I've force-pushed an updated version.
- rename `notify_on_exit()` to `notify_on_cleanup()`
- add `notify_start()`.

PTAL.

File: src/libsystemd/sd-daemon/daemon-util.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+                (void) sd_notify(false, start);
+
+        return stop;
+}
+
+/* This is intended to be used with _cleanup_ attribute. */
+static inline void notify_on_cleanup(const char **p) {
+        if (p)
+                (void) sd_notify(false, *p);
+}
```
Comment (keszybz) [inline src/libsystemd/sd-daemon/daemon-util.h]: This file should not be part of `sd-daemon`, unless you want to expose those helper functions as part of the public API at some point. But then they'd have to be defined in `sd-daemon.h` anyway. Please move this to `shared/`.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. Now `daemon-util.h` is moved to `src/shared/`. PTAL.

**Pull request merged**